CHANGELOG.md

```diff
@@ -28,6 +28,25 @@ PR [\#8658](https://github.com/decidim/decidim/pull/8658) Blocked users are pres
+
+For most cases, no changes in the code should be needed but there are specific differences with the implementation which may require changes in the 3rd party code as well. Both `Rectify::Form` and `Virtus` will be still available in the core (through the `rectify` gem) but you should migrate away from them as soon as possible as they may be removed in future versions of Decidim.
+
+There are specific things that you need to change regarding your Form or `Virtus.model` classes when migrating to `Decidim::AttributeObject`:
+
+- Change all instances of `YourForm < Rectify::Form` to `YourForm < Decidim::Form`. It should be very rare to find any classes in your code that inherit directly from `Rectify::Form` but in case you have used that, replace those references with `Decidim::Form`.
+- Change all instances `include Virtus.model` to `include Decidim::AttributeObject::Model`.
+- Change all `attribute :attr_name, Hash` to `attribute :attr_name, Hash[Symbol => ExpectedType]` where `ExpectedType` is the type you are expecting the hash values to be. The new layer will default the hash key types to `Symbol` and hash value types to `Object` (= any type). The Virtus Hash attribute did not force any default types for these. It should be preferred to use the actual expected type for the values instead of `Object` (= any type) to make your code more robutst and less buggy.
+- Change all `attribute :attr_name, Array` to `attribute :attr_name, Array[ExpectedType]` where `ExpectedType` is the type you are expecting the array values to be. It should be preferred to use the actual expected type for the values instead of `Object` (= any type) to make your code more robutst and less buggy.
+- The original form attribute values are no longer available through the `@attr_name` instance variables within the Form or `Virtus.model` classes. Instead, change all these references to `@attributes["attr_name"].value` in case you want to fetch the original value of the attribute without using its accessor method. Another way is to provide an alias for the original attribute method before overriding it. If you have not overridden the original attribute accessor, simply remove the `@` character in front of the attribute name to fetch the attribute value using the original accessor method.
+- When calling the `attributes` method of the model/form classes, use strings to refer to the attribute names, not symbols as you might have done with `Virtus` or `Rectify::Form`. Change all `model.attributes[:attr_name]` method calls to `model.attributes["attr_name"]`.
```

Comment:
This is something I was also thinking before. Some versions in the implementation history did just that but I feel that when we are migrating to `ActiveModel::Attributes`, we should respect their APIs instead of the Virtus APIs.

By default `ActiveModel::Attributes` only responds to the strings hash keys.

We could "fix" that by defining an override to the `attributes` method as `def attributes ; super.with_indifferent_access ; end` but I would see this quite hacky solution as going further I think we should match the API that is provided by `ActiveModel::Attributes`. It might also have some small performance impacts.

I was also thinking of adding a migration configuration (disabled by default) which would conditionally add this explained override to the `attributes` method but I am strongly adhering towards the API we are relying on under the middle-layer.

This is also why I mentioned in the CHANGELOG comments that this "aims to provide as much backwards compatibility as possible". This is one place where I would see the need for change honestly, although I completely understand your point of view.

In real life I think most of the times the required changes should be in specs, as in the forms, commands and controllers, the programmers should always prefer to call the accessor methods.

Reply:
@ahukkanen, i would go for 1-2 releases with the override mention, and a deprecation warning. 

```
def attributes 
  super.with_indifferent_access
end
```
This PR may kill most of the community based contributions like : 
- [Decidim Awesome](https://github.com/Platoniq/decidim-module-decidim_awesome)
- [NavBar Links](https://github.com/OpenSourcePolitics/decidim-module-navbar_links)

Reply:
@alecslupu I have made this change. This will mean that 3rd party developers will not update their code and will rely on the symbol accessors as long as it works fine, although it is not compatible with the underlying API.

Regarding the community based 3rd party modules, could you specify more clearly what you are expecting to break in these modules so that I can test it? By quickly glancing I wouldn't expect anything to break in the NavbarLinks module and in Decidim Awesome I think most of the things will work just fine without changes.

As mentioned before, mostly the changes are required in the specs of the 3rd party modules (change `attributes` hash expectation keys from symbols to strings). Other than that, it should be mostly backwards compatible now even with the symbol accessors after the latest change.

Other way you could fix the breaking specs is to change `expect(obj.attributes).to eq(foo: "bar")` into `expect(obj.to_h).to eq(foo: "bar")`. The `to_h` method will return a hash with symbol keys.

Reply:
@ahukkanen, i will give it a spin and i would let you know if my assumptions are validated.

Reply:
The current solution is great for me. I'll add an issue to remember to remove this hack in v0.30.